### ISSUE TITLE
Regression - less does not read whole pipe when stdin is /dev/null

### ISSUE THREAD

**Issue by mhlavink** [gwsw/less#558]
Hi,

we found and issue that is related to commit 953257f9030d29dd930ed3e76fd5811681b672b1 that was identified using bisect. 

less does not read whole piped input when stdin is /dev/null

We don't update all packages that often, so we've noticed this issue just now, when I've updated less and tests started to fail. I've tried to search for this issue report and found some similar (mostly old) issues, but this seems different.

Reproducer:
1) prepare zip file
```
$ date >file
$ zip file.zip file
```

2) run
`$ less file.zip </dev/null | cat`

Actual result: 
only partial output, one line mostly when run plain, two or two and half when run with strace:
```
Archive:  file.zip
```

Expected result:
```
Archive:  file.zip
Zip file size: 190 bytes, number of entries: 1
-rw-r--r--  3.0 unx       32 tx stor 24-Aug-09 05:25 file
1 file, 32 bytes uncompressed, 32 bytes compressed:  0.0%
``` 

tail of strace output used latest git:
```
readlink("/home/tester/less/file.zip", 0x7fff4e3771e0, 1023) = -1 EINVAL (Invalid argument)
pipe2([3<pipe:[22383]>, 4<pipe:[22383]>], O_CLOEXEC) = 0
prlimit64(0, RLIMIT_NOFILE, NULL, {rlim_cur=1024, rlim_max=512*1024}) = 0
prlimit64(0, RLIMIT_NOFILE, NULL, {rlim_cur=1024, rlim_max=512*1024}) = 0
mmap(NULL, 36864, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_STACK, -1, 0) = 0x7f631ebad000
rt_sigprocmask(SIG_BLOCK, ~[], [], 8)   = 0
clone3({flags=CLONE_VM|CLONE_VFORK|CLONE_CLEAR_SIGHAND, exit_signal=SIGCHLD, stack=0x7f631ebad000, stack_size=0x9000}, 88) = 19582
munmap(0x7f631ebad000, 36864)           = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
close(4<pipe:[22383]>)                  = 0
fcntl(3<pipe:[22383]>, F_SETFD, 0)      = 0
read(3<pipe:[22383]>, "A", 1)           = 1
ioctl(3<pipe:[22383]>, TCGETS, 0x7fff4e3784a0) = -1 ENOTTY (Inappropriate ioctl for device)
fstat(3<pipe:[22383]>, {st_mode=S_IFIFO|0600, st_size=0, ...}) = 0
write(1<pipe:[22370]>, "A", 1)          = 1
read(3<pipe:[22383]>, "rchive:  file.zip\nZip file size: 190 bytes, number of entries: 1\n", 8191) = 65
write(1<pipe:[22370]>, "rchive:  file.zip\nZip file size: 190 bytes, number of entries: 1\n", 65) = 65
poll([{fd=3<pipe:[22383]>, events=POLLIN}, {fd=0</dev/null>, events=POLLIN}], 2, 4000) = 1 ([{fd=0, revents=POLLIN}])
read(0</dev/null>, "", 1)               = 0
read(0</dev/null>, "", 1)               = 0
read(0</dev/null>, "", 1)               = 0
read(0</dev/null>, "", 1)               = 0
read(0</dev/null>, "", 1)               = 0
read(0</dev/null>, "", 1)               = 0
read(0</dev/null>, "", 1)               = 0
read(0</dev/null>, "", 1)               = 0
read(0</dev/null>, "", 1)               = 0
read(0</dev/null>, "", 1)               = 0
read(0</dev/null>, "", 1)               = 0
read(0</dev/null>, "", 1)               = 0
read(0</dev/null>, "", 1)               = 0
read(0</dev/null>, "", 1)               = 0
read(0</dev/null>, "", 1)               = 0
read(0</dev/null>, "", 1)               = 0
read(0</dev/null>, "", 1)               = 0
read(0</dev/null>, "", 1)               = 0
read(0</dev/null>, "", 1)               = 0
read(0</dev/null>, "", 1)               = 0
read(0</dev/null>, "", 1)               = 0
close(3<pipe:[22383]>)                  = 0
wait4(19582, [{WIFEXITED(s) && WEXITSTATUS(s) == 0}], 0, NULL) = 19582
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=19582, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
exit_group(1)                           = ?
+++ exited with 1 +++
```

**Comment by gwsw** [comment]
I can't reproduce this. One issue is I don't know exactly what your LESSOPEN script is doing. Setting ```LESSOPEN=|unzip -c %s``` doesn't reproduce it for me. 

However your strace output shows something that I think is more relevant: poll() is being called with ```poller[1].fd == 0```. This means ```tty==0```. I don't see how that can happen. tty is set in ```open_tty()```. First it tries ```ttyname(2)```, then ```"/dev/tty"```, then fd 2. Even if your system doesn't support ttyname and/or doesn't have /dev/tty, I don't see how any of these cases can result in tty being set to 0. Would it be possible for you to instrument open_tty and see where it sets tty?

**Comment by mhlavink** [comment]
`LESSOPEN=||/usr/bin/lesspipe.sh %s` and lesspipe.sh does `zipinfo -- $1`

I'm testing this on RHEL 9 (kernel-5.14.0-540, glibc-2.34-117), can be also reproduced on Fedora 39
System does support ttyname, check from configure:
```
...
checking for ttyname... yes
...
```
/dev/tty also exists

Not sure if this is what you wanted, but I tried this:
```
#include "less.h"

int main()
{
        int x=open_tty();
        printf("open_tty()=%d\n", x);
        printf("ttyname(%d)=%s\n",x,  ttyname(x));
        printf("ttyname(2)=%s\n", ttyname(2));
        return 0;
}
```
`$ gcc test.c -I. ttyin.o -Wall -o otty`
```
$ ./otty
open_tty()=3
ttyname(3)=/dev/tty
ttyname(2)=/dev/pts/0
```

**Comment by mhlavink** [comment]
this is strace log created with this:

`$ strace -f -o less-strace.log -y -s 100 less file.zip </dev/null | cat`

[less-strace.log](https://github.com/user-attachments/files/16567349/less-strace.log)

with all parent and children, plus less only log filtered from same file with

`$ sed -n -e 's|^4715  ||p' <less-strace.log >less-only-strace.log` 

[less-only-strace.log](https://github.com/user-attachments/files/16567348/less-only-strace.log)

**Comment by gwsw** [comment]
Thanks. I think I see the problem. Can you confirm whether this patch fixes it for you? This is a patch against less-661; if you are using a different version let me know and I can provide a patch for it.

```
diff --git a/os.c b/os.c
index 61c336c..9f89777 100644
--- a/os.c
+++ b/os.c
@@ -80,6 +80,7 @@ extern int exit_F_on_close;
 extern int follow_mode;
 extern int scanning_eof;
 extern char intr_char;
+extern int is_tty;
 #if !MSDOS_COMPILER
 extern int tty;
 #endif
@@ -237,7 +238,7 @@ start:
    }
 #endif
 #if USE_POLL
-   if (fd != tty && use_poll)
+   if (is_tty && fd != tty && use_poll)
    {
        int ret = check_poll(fd, tty);
        if (ret != 0)
```

**Comment by mhlavink** [comment]
Thanks for quick patch. I've tested it and run also other tests and it works now

**Comment by gwsw** [comment]
Fixed in ed454a217da417dc052723ea70da8efde0f2e66c.
